pkg/ebpf/c/offset-guess.c

```diff
@@ -0,0 +1,185 @@
+#include <linux/kconfig.h>
+#include <uapi/linux/ptrace.h>
+#include "bpf_helpers.h"
+#include "offset-guess.h"
+
+#include <net/sock.h>
+#include <net/net_namespace.h>
+#include <uapi/linux/tcp.h>
+
+/* These maps are used to match the kprobe & kretprobe of connect for IPv6 */
+/* This is a key/value store with the keys being a pid
+ * and the values being a struct sock *.
+ */
+struct bpf_map_def SEC("maps/connectsock_ipv6") connectsock_ipv6 = {
+    .type = BPF_MAP_TYPE_HASH,
+    .key_size = sizeof(__u64),
+    .value_size = sizeof(void*),
+    .max_entries = 1024,
+    .pinning = 0,
+    .namespace = "",
+};
+
+struct bpf_map_def SEC("maps/tracer_status") tracer_status = {
+    .type = BPF_MAP_TYPE_HASH,
+    .key_size = sizeof(__u64),
+    .value_size = sizeof(tracer_status_t),
+    .max_entries = 1,
+    .pinning = 0,
+    .namespace = "",
+};
+
+__attribute__((always_inline))
+static bool proc_t_comm_equals(proc_t a, proc_t b) {
+    int i;
+    for (i = 0; i < TASK_COMM_LEN; i++) {
+        if (a.comm[i] != b.comm[i]) {
+            return false;
+        }
+    }
+    return true;
+}
+
+__attribute__((always_inline))
+static bool check_family(struct sock* sk, tracer_status_t* status, u16 expected_family) {
+    u16 family = 0;
+    bpf_probe_read(&family, sizeof(u16), ((char*)sk) + status->offset_family);
+    return family == expected_family;
+}
+
+__attribute__((always_inline))
+static int guess_offsets(tracer_status_t* status, struct sock* skp) {
+    u64 zero = 0;
+
+    if (status->state != TRACER_STATE_CHECKING) {
+        return 1;
+    }
```

Comment:
I don't think this much matters, but do we still need this check now that this kprobe is removed once offset guessing is done?

Reply:
I think this is still required. Otherwise if the probe fires while userspace is figuring out if offset guessing is complete or not, it could do unnecessary/unwanted work.